doc/guide/transactions-and-threading.rst

```diff
@@ -0,0 +1,394 @@
+data structures, like indexes, so the updates can be made by a
+dedicated thread or process.
+
+Conflict resolution
+~~~~~~~~~~~~~~~~~~~
+
+ZODB provides a conflict-resolution framework for merging conflicting
+changes. This is implemented by `BTree
+<https://pythonhosted.org/BTrees/>`_ buckets and ``Length`` objects.
+
+The main data structures provided by BTrees: BTrees and TreeSets,
+spread their data over multiple objects.  The leaf-level objects,
+called *buckets* allow distinct keys to be updated without causing
+conflicts [#usually-avoids-conflicts]_.
+
+``Length`` objects are conflict key counters, that merge changes by
+simply accumulating changes.
+
+The use of BTree buckets, and to a lesser degree ``Length`` objects
+is a very common technique.
+
+.. caution::
+   Conflict resolution weakens consistency.  Resist the temptation to
+   try to implement conflict resolution yourself.  In the future, ZODB
+   will provide greater control over conflict resolution, including
+   the option of disabling it.
+
+   It's generally best to avoid conflicts in the first place, if possible.
+
+ZODB and atomicity
+==================
+
+ZODB provides atomic transactions. When using ZODB, it's important to
+align work with transactions.  Once a transaction is committed, it
+can't be rolled back [#undo]_ automatically.  For applications, this
+implies that work that should be atomic shouldn't be split over
+multiple transactions.  This may seem somewhat obvious, but the rule
+can be broken in non-obvious ways. For example a Web API that splits
+logical operations over multiple web requests, as is often done in
+`REST
+<https://en.wikipedia.org/wiki/Representational_state_transfer>`_
+APIs, violate this rule.
+
+Partial transaction error recovery using savepoints
+---------------------------------------------------
+
+A transaction can be split into multiple steps that can be rolled back
+individually.  This is done by creating savepoints.  Changes in a
+savepoint can be rolled back without rolling back an entire
+transaction::
+
+  import ZODB
+  db = ZODB.DB(None) # using a mapping storage
+  with db.transaction() as conn:
+      conn.root.x = 1
+      conn.root.y = 0
+      savepoint = conn.transaction_manager.savepoint()
+      conn.root.y = 2
+      savepoint.rollback()
+
+  with db.transaction() as conn:
+      print([conn.root.x, conn.root.y]) # prints 1 0
+
+.. -> src
+
+   >>> exec(src)
+   [1, 0]
+
+If we executed this code, it would print 1 and 0, because while the
+initial changes were committed, the changes in the savepoint were
+rolled back.
+
+A secondary benefit of savepoints is that they save any changes made
+before the savepoint to a file, so that memory of changed objects can
+be freed if they aren't used later in the transaction.
+
+Concurrency, threads and processes
+==================================
+
+ZODB supports concurrency through transactions.  Multiple programs can
```

Comment:
programs -> tasks?

programs is an overloaded word, many people might think of operating system processes.